### PR TITLE
Add background foreground

### PR DESCRIPTION
**Pull request by hokyungh**
### Issue #, if available:
Chime-33481
### Description of changes:
Add logic of stopping local video 
### Testing done:
Screen Shot 2020-12-15 at 4 57 30 PM

Screen Shot 2020-12-15 at 4 58 05 PM

1. Background and foreground the app. In the background, confirm that network usage is low and local video is no longer sending
2. Pause the video and background/foreground the app. Confirm that it still shows paused video
3. While backgrounded, turn the one of remote video off. Confirm that it is removed when foregrounded.
4. While backgrounded, turn the one of remote video on. Confirm that it is added when foregrounded.
5. Join as Callkit. Video behaves the same.

#### Manual test cases (add more as needed):

- [X] Join meeting without CallKit
- [x] Join meeting with CallKit as Incoming
- [x] Join meeting with CallKit as Outgoing
- [X] Leave meeting
- [X] Rejoin meeting
- [X] See metrics
- [X] See attendee presence
- [X] Send audio
- [X] Receive audio
- [X] Mute self
- [X] Unmute self
- [X] See local mute indicator when muted
- [X] See remote mute indicator when other is muted
- [X] Enable and disable local video
- [X] Enable and disable remote video from remote side
- [X] Send local video
- [X] Pause and resume remote video
- [X] Switch local camera
- [ ] Receive remote screen share

#### Integration validation (required before release)

- [ ] Unit tests pass
- [ ] Build SDK against simulator with release options
- [ ] Build SDK against device with release options
- [ ] Build SDK against simulator with debug options
- [ ] Build SDK against device with debug options
- [ ] Test Demo against simulator with SDK (debug build) in workspace
- [ ] Test Demo against device with SDK (debug build) in workspace
- [ ] Test DemoObjC against simulator with SDK (debug build) in workspace
- [ ] Test DemoObjC against device with SDK (debug build) in workspace
- [ ] Test Demo against simulator with SDK.framework (release build)
- [ ] Test Demo against device with SDK.framework (release build)
- [ ] Test DemoObjC against simulator with SDK.framework (release build)
- [ ] Test DemoObjC against device with SDK.framework (release build)

### Screenshots, if available:

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 ## Unreleased
-
+### Added
+* Update demo application to pause/resume when app is in background/foreground
```
Comment (tianyux-amz) [inline CHANGELOG.md:3]: Nit: Blank line after this line?

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift [hunk 2]
```diff
@@ -58,6 +58,9 @@ class MeetingViewController: UIViewController {
 
     // Model
     var meetingModel: MeetingModel?
+    
+    // isLocalOn
+    private var isLocalOn: Bool = false
```
Comment (tianyux-amz) [inline AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift:63]: Nit: rename to `wasLocalVideoOn`? Also, do not just repeat the variable name as a comment 😊

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift [hunk 3]
```diff
@@ -88,6 +91,31 @@ class MeetingViewController: UIViewController {
...
+        NotificationCenter.default.addObserver(self, selector: #selector(appMovedToForeground), name: UIApplication.willEnterForegroundNotification, object: nil)
+    }
+    
+    private func deregisterForAppLifeCycleNotification() {
+        NotificationCenter.default.removeObserver(self, name: UIApplication.didEnterBackgroundNotification, object: nil)
+        NotificationCenter.default.removeObserver(self, name: UIApplication.willEnterForegroundNotification, object: nil)
+    }
+    
+    @objc private func appMovedToBackground() {
+        isLocalOn = meetingModel?.isLocalVideoActive ?? false
```
Comment (tianyux-amz) [inline AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift:106]: Can we just do `isLocalOn = meetingModel?.isLocalVideoActive`?

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift [hunk 4]
```diff
@@ -371,6 +400,7 @@ class MeetingViewController: UIViewController {
     @IBAction func leaveButtonClicked(_: UIButton) {
         meetingModel?.endMeeting()
         deregisterFromKeyboardNotifications()
+        deregisterForAppLifeCycleNotification()
```
Comment (tianyux-amz) [inline AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift:403]: Will these be called if we leave meeting from CallKit?

**Review by zhinang-amazon — COMMENTED** [review]
All the code added in MeetingViewController does not seem to be view related.
They should all be moved into MeetingModel.
And make sure CallKit Incoming and Outgoing are also handled.

App life cycle events are normally handled in `AppDelegate` callbacks. Even though registering for notifications from MeetingModel works for our app, but for a more complex app, they maybe saving multiple MeetingModel instances. 
I would prefer to see that in AppDelegate [applicationDidEnterBackground](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622997-applicationdidenterbackground) and [applicationWillEnterForeground](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623076-applicationwillenterforeground) callback, access `MeetingModule.shared().activeMeeting` to do the corresponding actions.

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift [hunk 5]
```diff
@@ -88,6 +91,31 @@ class MeetingViewController: UIViewController {
...
+    
+    @objc private func appMovedToBackground() {
+        isLocalOn = meetingModel?.isLocalVideoActive ?? false
+        if isLocalOn {
+            meetingModel?.isLocalVideoActive = false
+        }
+        meetingModel?.videoModel.pauseAllRemoteVideos()
+    }
+    
+    @objc private func appMovedToForeground() {
```
[the remark this answers is not in the document]
Comment (hokyungh) [reply AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift]: removed

**Comment by hokyungh** [comment]
> All the code added in MeetingViewController does not seem to be view related.
> They should all be moved into MeetingModel.
> And make sure CallKit Incoming and Outgoing are also handled.
> 
> App life cycle events are normally handled in `AppDelegate` callbacks. Even though registering for notifications from MeetingModel works for our app, but for a more complex app, they maybe saving multiple MeetingModel instances.
> I would prefer to see that in AppDelegate [applicationDidEnterBackground](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622997-applicationdidenterbackground) and [applicationWillEnterForeground](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623076-applicationwillenterforeground) callback, access `MeetingModule.shared().activeMeeting` to do the corresponding actions.

Updated accordingly

File: CHANGELOG.md (hunk 1 above)
Comment (hokyungh) [reply CHANGELOG.md:3]: Added

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift (hunk 2 above)
Comment (hokyungh) [reply AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift:63]: Updated the name and moved to `MeetingModel`

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift (hunk 3 above)
Comment (hokyungh) [reply AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift]: updated

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift (hunk 4 above)
Comment (hokyungh) [reply AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift]: See the tests on the description

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift (hunk 2 above)
Comment (zhinang-amazon) [reply AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift:63]: Forgot to delete this?

**Review by zhinang-amazon — DISMISSED** [review]

File: AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift [hunk 6]
```diff
@@ -58,7 +58,7 @@ class MeetingViewController: UIViewController {
 
     // Model
     var meetingModel: MeetingModel?
-
+
```
Comment (tianyux-amz) [inline AmazonChimeSDKDemo/AmazonChimeSDKDemo/MeetingViewController.swift:61]: Remove this?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,7 @@
+## Unreleased
+### Added
```
Comment (tianyux-amz) [inline CHANGELOG.md:2]: Blank line before this?

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,7 @@
...
+* Update demo application to pause/resume when app is in background/foreground
```
Comment (tianyux-amz) [inline CHANGELOG.md:3]: "pause/resume" -> "pause/resume local and remote videos"?

**Review by tianyux-amz — APPROVED** [review]

**Review by zhinang-amazon — APPROVED** [review]

**Pull request merged**